switchboard/tests/test_manager.py

```diff
@@ -735,3 +735,30 @@ def verify_context():
 
         if self.exc:
             raise self.exc
+
+class TestManagerResultCaching(object):
```

Comment:
This strikes me as a higher-level integration test of result caching, which is good... along the same lines of the TestAPI class. I'd also like to see something lower level that focuses much more tightly on the decorator, with the intent of testing all the logic branches within (including raising a TypeError):

~~~~
def test_with_result_cache_disabled(self):
    # By default caching is off, so result_cache is None.
    foo = 'bar'

    @test_with_result_cache
    def tester():
        return foo

    assert_equals(tester(), 'bar')
    foo = 'baz'
    assert_equals(tester(), 'baz')
~~~~